File: python/stack_queue/Solution735.py
# 735. 小行星碰撞
from typing import List


class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []
        for a in asteroids:
            while stack and a < 0 < stack[-1]:
                # collision happens
                crasher = stack[-1]
                if a + crasher == 0:
                    stack.pop()
                elif a + crasher < 0:
                    stack.pop()
                    continue
                break
            else:
                stack.append(a)
        return stack

File: src/stack_queue/Solution735.java
// 735. 小行星碰撞
package stack_queue;

import java.util.Deque;
import java.util.LinkedList;

public class Solution735 {

    public int[] asteroidCollision(int[] asteroids) {
        Deque<Integer> stack = new LinkedList<>();
        for (int a: asteroids) {
            boolean crashed = false;
            while (!stack.isEmpty() && a < 0 && 0 < stack.peek()) {
                int crasher = stack.peek();
                if (crasher + a > 0) {
                    crashed = true;
                    break;
                } else if (crasher + a == 0) {
                    stack.pop();
                    crashed = true;
                    break;
                } else {
                    stack.pop();
                }
            }
            if (!crashed) stack.push(a);
        }
        int[] ans = new int[stack.size()];

        int j = stack.size() - 1;
        while (j >= 0) {
            ans[j] = stack.pop();
            j--;
        }
        return ans;
    }

    public static void main(String[] args) {
        int[] ast = new int[]{5, 10, -5};
        Solution735 sol = new Solution735();
        int[] ans = sol.asteroidCollision(ast);
        for (Integer a: ans) {
            System.out.println(a);
        }
    }
}
